### ISSUE TITLE
Is it possible to parepare two set of weights for feedforward and backprop separately?

### ISSUE THREAD

**Issue by developer-mayuan** [tensorflow/tensorflow#19028]
Currently I'm working on the network quantization and I try to implement the ideas of the paper ["Mixed precision training"](https://arxiv.org/abs/1710.03740) using tensorflow. In this paper, the author prepare two set of weights (with different data type) for feedforward and backprop separately. I want to know if current tensorflow can support this kind of feature. Thanks.

![screenshot from 2018-05-02 10-14-31](https://user-images.githubusercontent.com/13360081/39538427-a48b88fa-4df1-11e8-99ee-e2ee284b94a9.png)

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
TensorFlow version
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce

**Comment by asimshankar** [comment]
A GitHub issue is probably not the best forum for this, but @zheng-xq @protoget @bignamehyp may want to comment as they are looking into making this easier going forward. But yes, it is possible right now (by keeping variables as floats, but casting to half precision before computation, and then scaling the loss/gradients before applying them), and they may want to chime in with details.

**Comment by protoget** [comment]
It isn't really two sets of weights. The weights are casted and used in following computation. The gradients are casted back to variable dtype before applied on weights.

We're working on supporting it at tensorflow layer level for mixed precision training. Its goal is to support training with ops done in one dtype and variables in a separate dtype, which seems to solve most of our usecases so far.
